### DISCUSSION TITLE
Why does my compilation fail?

### DISCUSSION THREAD

**Discussion by DeafMan1983** [unoplatform/Uno.Wasm.Bootstrap#513] [Q&A]
Hello everyone,

I am using Ubuntu 22.04 with dotnet 6 sdk 
And why does it throw if I have cloned emsdk from github and it said my emscription is unsupported? But my emscripten has very latest version.

```
Microsoft (R) Build Engine version 17.2.0+41abc5629 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  MyWebConsole -> /home/deafman1983/CSharpDevelopments/MyWebConsole/bin/Debug/net6.0/MyWebConsole.dll
/home/deafman1983/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/Uno.Wasm.Bootstrap.targets(175,5): error : [00:00:00.0064181] /home/deafman1983/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/scripts/emscripten-setup.sh: line 13: mono: command not found [/home/deafman1983/CSharpDevelopments/MyWebConsole/MyWebConsole.csproj]
/home/deafman1983/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/Uno.Wasm.Bootstrap.targets(175,5): error : System.NotSupportedException: Failed to install emscripten [/home/deafman1983/CSharpDevelopments/MyWebConsole/MyWebConsole.csproj]
/home/deafman1983/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/Uno.Wasm.Bootstrap.targets(175,5): error :    at Uno.Wasm.Bootstrap.ShellTask_vfc89b40db92a14b8f04d4c5c9c72690bdcd7b878.ValidateEmscripten() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 1106 [/home/deafman1983/CSharpDevelopments/MyWebConsole/MyWebConsole.csproj]
/home/deafman1983/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/Uno.Wasm.Bootstrap.targets(175,5): error :    at Uno.Wasm.Bootstrap.ShellTask_vfc89b40db92a14b8f04d4c5c9c72690bdcd7b878.RunPackager() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 728 [/home/deafman1983/CSharpDevelopments/MyWebConsole/MyWebConsole.csproj]
/home/deafman1983/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/Uno.Wasm.Bootstrap.targets(175,5): error :    at Uno.Wasm.Bootstrap.ShellTask_vfc89b40db92a14b8f04d4c5c9c72690bdcd7b878.Execute() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 228 [/home/deafman1983/CSharpDevelopments/MyWebConsole/MyWebConsole.csproj]
```

Why do I need mono? But mono is died on Ubuntu 22.04 since I have tested with Godot Mono and it looks unusable that is why I don't want install Mono runtime. I want use only dotnet 6.0.5x

But why do you need mono runtime? Please stop using Mono Runtime! Mono Runtime should bury since latest version of Ubuntu. 

Please use only normal with Maui and Blazor-WASM!

I tried to explain you:
I create myrustlib like [this](https://platform.uno/blog/c-interop-with-c-c-and-rust-in-webassembly/).

And type `dotnet restore` and `dotnet build`

Thank you!

**Comment by jeromelaban** [comment]
The documentation is not exactly up to date, but you'll need to [install mono and ninja](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/runtime-execution-modes.md#required-configuration-for-aot-mixed-mode-or-static-linking-on-linux). The mono link is in the WSL section, but it applies to "native" Linux as well.

**Comment by DeafMan1983** [comment]
WSL? Sorry I use native Ubuntu 22.04. But I have compiled since Godot with Mono but it crashed because Mono is already dead that's why what is with blazor wasm? It is very new template of Dotnet 6.0 workloads `wasm-tools`.

What does it happen with bc files?

Blazor WASM:
Via nativereference with C files  like example 
libSDL.c 
```
#include <SDL/SDL.h>
// it builds for emscripten
```
Then Dotnet builds single wasm file and check browser.

But why do I need to install Mono Runtime? It is really impossible to support for dead Runtime Mono. We need leave Mono and use only Dotnet.

**Reply by jeromelaban** [reply]
The requirement of mono is just an artifact of the build, it will be removed in the future. The rest of the build is using dotnet 5 or 6.

I updated [the documentation here](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/runtime-execution-modes.md#required-configuration-for-aot-mixed-mode-or-static-linking-on-linux). I mentioned WSL because the steps were the right one before I changed the documentation.

**Reply by DeafMan1983** [reply]
Great news - I have tried to use blazorwasm with bc files = It works fine like you made old way :D 

I resolved like this
You need add `<NativeFileReference Include="myrustlib.bc" />`
in blazorwasm generated project with `<ItemGroup> ... </ItemGroup>`

I replace `Program.cs`
```
using Microsoft.AspNetCore.Components.Web;
using Microsoft.AspNetCore.Components.WebAssembly.Hosting;

using System.Runtime.InteropServices;

namespace MyWASM;

class Program
{
    const string myrust = "myrustlib";
    [DllImport(myrust)]
    static extern int rust_add(int a, int b);

    static int Main(string[] args)
    {
        Console.WriteLine("Sum is " + rust_add(5, 5));
        return 0;
    }
}
```

`dotnet build` and it builds successfully :+1: 
dotnet run and copy -past or execute link to open browser and open Development tool:
![image](https://user-images.githubusercontent.com/57066679/171978445-e261df41-9a30-41cc-8fa5-0886b264aaa6.png)
Viola!!! I am surprised because bc embeds in wasm of dotnet thanks Microsoft!!!!!

I expect that is future version "Blazor WASM" with bc files = Successfully! Without dead Mono Runtime.

Can you test with dotnet new blazorwasm -o my_wasm, cd my_wasm and code . and add path to bc file and build with dotnet and result = It is successful working. 

Please add my credit to your documentation for blazorwasm project ( great way without mono runtime )

**Reply by jeromelaban** [reply]
Do you mean this [existing documentation](https://docs.microsoft.com/en-us/aspnet/core/blazor/webassembly-native-dependencies?view=aspnetcore-6.0)?

If you want to use .NET blazor directly, go ahead, but don't be disrespectful.

**Reply by DeafMan1983** [reply]
Don't worry! I will be respectfully. I want improve about unoplatform and blazorwasm because unoplatform was initial time since they create webassembly from dotnet with Mono Runtime. Now blazorwasm has own emsdk installation ( wasm-tools ). Thanks and saving my suggestion with blazorwasm!

Happy Pentecost Weekend!
